### PR TITLE
feat(pulumi): add pulumi package to handle 1.0 deployments

### PR DESCRIPTION
**Pull request by vanbasten17**
## Description
* Added pulumi classes to handle deploying with pulumi in a headless way and Botonic AWS Components to have all the infrastructure uploaded to AWS.
* **Note:** 1st approach, some parts are susceptible to change.

## Approach taken / Explain the design
`PulumiRunner` is the main class to handle all the logic (it is the one that will be used within the deploy script, probably in @botonic/cli). Given the config of the user, which can include the developer aws credentials, it will deploy the full-stack in two steps (due to some limitations with Pulumi, we couldn't handle to do it with a single component).

1. Run build commands for websocket server, rest server and webchat.

2. Deploying backend
* Deployment of DynamoDB database.
* Deployment of NLP models in a static s3 Bucket.
* Deployment of WebSocket Server (it will use previously generated database url and nlp models bucket url in order to fill the variables of their lambda functions. This deployment will deploy three separate lambdas in order to handle the `onConnect`, `onMessage` and `onDisconnect` events.
* Deployment of Rest Server (will use also database url and nlp models bucket url). It packages the express server app into a single lambda function which will be used to process REST requests.

**(parentheses)**: after the backend has been deployed, the websocket url is used to replace the variable referencing the websocket endpoint (in webchat.botonic.js) where webchat needs to call the ws-api.

3. Deploying frontend
* Deploy static contents (webchat + assets)
* Deploy the CDN with a Cloudfront distribution (and make use of the previous resources endpoints to create aliases for every api endpoint( websocket and rest))

## To document / Usage example
**e.g.:**
```
const websocketBuildCmd = {
  name: 'WebSocket Server Build',
  command: 'yarn workspace api build:websocket',
}

const restServerBuildCmd = {
  name: 'Rest Server Build',
  command: 'yarn workspace api build:rest',
}

const staticBuildCmd = {
  name: 'Static Contents Build',
  command: 'yarn workspace webchat build',
}

const runner = new PulumiRunner(
  [websocketBuildCmd, restServerBuildCmd, staticBuildCmd],
  {
    projectName: 'myProject',
    stackName: 'development',
    region: 'eu-west-1',
    profile: 'my-aws-profile',
    tags: {
      Scope: 'Botonic-1.0',
      Maintainer: 'team.botonic',
    },
    tableName: 'user-events',
    customDomain: 'myBotName.myBusiness.com'
  }
).run()
```
this will deploy the entire stack given the following output once finished:
```
// Outputs:
Assets --> https://myBotName.myBusiness.com/
Rest Api --> https://myBotName.myBusiness.com/api/
WebSocket Api --> wss://myBotName.myBusiness.com/ws-api/
```
Then we could be able to retrieve our webchat script: `https://myBotName.myBusiness.com/webchat.botonic.js`, making an api call to our rest endpoint `curl https://myBotName.myBusiness.com/api/users/` or connect to our websocket api `wscat -c wss://myBotName.myBusiness.com/ws-api/`.

### CODE REVIEW

File: packages/botonic-pulumi/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+export const getNamePrefix = (): string => {
+  const config = new Config()
+  return `botonic-${config.get('projectName')}-${config.get('stackName')}`
+}
+export const NLP_MODELS_PATH = join(cwd(), 'bot', 'src', 'nlp', 'tasks')
+export const WEBSOCKET_SERVER_PATH = join(cwd(), 'api', 'dist', 'websocket')
+export const REST_SERVER_PATH = join(cwd(), 'api', 'dist', 'rest')
+export const WEBCHAT_CONTENTS_PATH = join(cwd(), 'webchat', 'dist')
+export const WEBSOCKET_ENDPOINT_PATH_NAME = 'ws-api'
```
Comment (ericmarcos) [inline packages/botonic-pulumi/src/index.ts:13]: Better just `ws`?

**Review by ericmarcos — APPROVED** [review]
Amazing stuff!! 👏👏👏👏

File: packages/botonic-pulumi/src/aws/index.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import * as aws from '@pulumi/aws'
+import * as pulumi from '@pulumi/pulumi'
+
+import { getNamePrefix } from '..'
+
+// DynamoDB
+export const DYNAMODB_TABLE_NAME = 'user_events'
```
[the remark this answers is not in the document]
Comment (vanbasten17) [reply packages/botonic-pulumi/src/aws/index.ts:7]: named it as `BOTONIC_SINGLE_TABLE_NAME`

[the remark this answers is not in the document]
Comment (vanbasten17) [reply packages/botonic-pulumi/src/aws/index.ts:7]: leave it as `BOTONIC_SINGLE_TABLE_NAME`

File: packages/botonic-pulumi/src/aws/deployment-stacks.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    nlpModelsBucketEndpoint: nlpModelsBucket.endpoint,
+    databaseEndpoint: database.endpoint,
+    websocketServerEndpoint: websocketServer.endpoint,
+    restServerEndpoint: restServer.endpoint,
+  }
+}
+
+interface FrontendDeployResults {
+  cloudFrontDomainEndpoint: pulumi.Output<string>
+}
```
[the remark this answers is not in the document]
Comment (vanbasten17) [reply packages/botonic-pulumi/src/aws/deployment-stacks.ts:63]: now in the frontend deploy outputting (no-custom):
```
Outputs:
    apiUrl      : "https://drlykjvqxni0g.cloudfront.net/api/"
    nlpModelsUrl: "https://drlykjvqxni0g.cloudfront.net/"
    webchatUrl  : "https://drlykjvqxni0g.cloudfront.net/"
    websocketUrl: "wss://drlykjvqxni0g.cloudfront.net/ws/"
```
and:
```
Outputs:
  apiUrl      : "https://bot-1.domain.com/api/"
  nlpModelsUrl: "https://bot-1.domain.com/"
  webchatUrl  : "https://bot-1.domain.com/"
  websocketUrl: "wss://bot-1.domain.com/ws/"
```

File: packages/botonic-pulumi/src/aws/deployment-stacks.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import { RestServer } from './rest-server'
+import { StaticWebchatContents } from './static-webchat-contents'
+import { WebSocketServer } from './websocket-server'
+
+interface BackendDeployResults {
+  nlpModelsBucketEndpoint: pulumi.Output<string>
+  databaseEndpoint: pulumi.Output<string>
+  websocketServerEndpoint: pulumi.Output<string>
+  restServerEndpoint: pulumi.Output<string>
+}
```
[the remark this answers is not in the document]
Comment (vanbasten17) [reply packages/botonic-pulumi/src/aws/deployment-stacks.ts]: done 👍

File: packages/botonic-pulumi/src/aws/index.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+// Websocket Server
+export const WEBSOCKET_LAMBDA_HANDLER_NAME = 'server.default'
+export const WEBSOCKET_ONCONNECT_LAMBDA_NAME = 'onConnect'
+export const WEBSOCKET_ONCONNECT_ROUTE_KEY = '$connect'
+export const WEBSOCKET_ONMESSAGE_LAMBDA_NAME = 'onMessage'
+export const WEBSOCKET_ONMESSAGE_ROUTE_KEY = '$default'
+export const WEBSOCKET_ONDISCONNECT_LAMBDA_NAME = 'onDisconnect'
+export const WEBSOCKET_ONDISCONNECT_ROUTE_KEY = '$disconnect'
```
[the remark this answers is not in the document]
Comment (vanbasten17) [reply packages/botonic-pulumi/src/aws/index.ts]: yes, I added here if for some reason we decided to change some of them, but maybe I went too far away, keeping it simpler

File: packages/botonic-pulumi/src/index.ts (hunk 1 above)
Comment (vanbasten17) [reply packages/botonic-pulumi/src/index.ts]: I agree 👍 , renamed

**Comment by vanbasten17** [comment]
feedback applied @ericmarcos

**Pull request merged**